container/container.go

```diff
@@ -530,9 +531,26 @@ func copyEscapable(dst io.Writer, src io.ReadCloser, keys []byte) (written int64
 // ShouldRestart decides whether the daemon should restart the container or not.
 // This is based on the container's restart policy.
 func (container *Container) ShouldRestart() bool {
-	return container.HostConfig.RestartPolicy.Name == "always" ||
-		(container.HostConfig.RestartPolicy.Name == "unless-stopped" && !container.HasBeenManuallyStopped) ||
-		(container.HostConfig.RestartPolicy.Name == "on-failure" && container.ExitCode != 0)
+	rp := container.HostConfig.RestartPolicy
+	switch {
+	case rp.IsAlways():
+		return true
+	case rp.IsUnlessStopped():
+		if !container.HasBeenManuallyStopped {
```

Comment:
It's a nit but I would have inverted the condition, so it's `if container.HasBeenManuallyStopped {` instead (feels easier to read).